### PR TITLE
Support UI Authentication for OpenID Connect accounts

### PR DESCRIPTION
**Pull request by clokep**
This is similar to #7102 / #7186, but for OpenID (added in #7256).

It also cleans up a little bit of code to make things consistent between the various SSO workflows.

CC @sandhose

### CODE REVIEW

**Comment by sandhose** [comment]
LGTM. Nothing really changed in the OIDC flow, just that the session cookie has one more key in it (and that part is covered enough by unit tests)

**Comment by clokep** [comment]
@sandhose Do you think the macaroon changes are reasonable? Sometimes having a particular key there and sometimes now. I was unsure if this is "valid", but it seems to pass verification OK still.

**Comment by sandhose** [comment]
> @sandhose Do you think the macaroon changes are reasonable? Sometimes having a particular key there and sometimes now. I was unsure if this is "valid", but it seems to pass verification OK still.

Wasn't sure either but I tested locally and it works. Seems like caveat verification in macaroons means that each first party caveat in the token must [satisfy at least one verifier](https://github.com/ecordell/pymacaroons/blob/06b55110eda2fb192c130dee0bcedf8b124d1056/pymacaroons/caveat_delegates/first_party.py#L39-L43). It does not mean that each verifier must be used.

File: synapse/handlers/oidc_handler.py [hunk 1]
```diff
@@ -497,7 +497,10 @@ async def _parse_id_token(self, token: Token, nonce: str) -> UserInfo:
         return UserInfo(claims)
 
     async def handle_redirect_request(
-        self, request: SynapseRequest, client_redirect_url: bytes
+        self,
+        request: SynapseRequest,
+        client_redirect_url: bytes,
+        ui_auth_session_id: Optional[str] = None,
     ) -> None:
```
Comment (richvdh) [inline synapse/handlers/oidc_handler.py:504]: this now returns a `str` rather than a None? could do with updating the docstring too.

Comment (clokep) [reply synapse/handlers/oidc_handler.py]: Good catch. Unfortunate mypy didn't see this. 😮

**Pull request merged**